dvc/remote/local/__init__.py

```diff
@@ -572,3 +572,45 @@ def _get_unpacked_dir_names(self, checksums):
             if self.is_dir_checksum(c):
                 unpacked.add(c + self.UNPACKED_DIR_SUFFIX)
         return unpacked
+
+    def _needs_checkout(self, path_info, checksum):
+        # NOTE: In case if path_info is already cached, and cache type is
+        # 'copy' we would like to avoid relinking.
+        # `symlink/hardlink/reflink` is cheap, but if we need `copy` links,
+        # then to save on copying we should only relink if we already have
+        # `hardlink` or a `symlink`, we don't care about reflinks, because
+        # they are indistinguishable from copy anyway.
+
+        if (
+            self._is_cache_copy(path_info)
+            and not self.changed(path_info, {self.PARAM_CHECKSUM: checksum})
+            and not System.is_hardlink(path_info)
+            and not System.is_symlink(path_info)
+        ):
+            msg = "File '{}' didn't change"
+            logger.debug(msg.format(str(path_info)))
+            return False
+        return True
+
+    def _is_cache_copy(self, path_info):
+        # NOTE: path_info required to make test reliable, when cache is on
+        # different fs than path_info
+
+        test_cache_file = self.path_info / ".cache_type_test_file"
```

Comment:
Move this lower, next to `workspace_file = ...`.

Reply:
done